opt: plan index joins for non-covering indexes

justinj opened the pull request:
This commit introduces the ability to explore index scans which are not
covering the requested columns, and use an index join to add in the
columns which are required.

It does not yet support execbuilding such plans, however, and they're
given infinite cost so they don't get planned.

This is also a good bit less general than the full machinery of planning lookup joins. 

Release note: None

Commit:
opt: plan index joins for non-covering indexes

This commit introduces the ability to explore index scans which are not
covering the requested columns, and use an index join to add in the
columns which are required.

It does not yet support execbuilding such plans, however, and they're
given infinite cost so they don't get planned.

This is also a good bit less general than the full machinery of planning
lookup joins.

Release note: None

RaduBerinde commented:
Just glanced over it for now, nice work overall!

Is the plan that later when we have LookupJoin, the IndexJoin op will go away? Or would we keep both?

---

Review status: 0 of 17 files reviewed at latest revision, all discussions resolved.

---

*[pkg/sql/opt/norm/testdata/combo, line 317 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/25019#-LAoQMaF3aGvmbKZDKYW:-LAoQMaF3aGvmbKZDKYX:bw7zewq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/norm/testdata/combo#L317)):*
> ```
>   - └── projections [outer=(4)]
>   -      └── variable: a.s [type=string, outer=(4)]
>   +unknown
> ```

unknown?

---

*[pkg/sql/opt/xform/explorer.go, line 243 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/25019#-LAoRPFQCH0EM29s9Qga:-LAoRPFQCH0EM29s9Qgb:bb8tpxm) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/xform/explorer.go#L243)):*
> ```Go
> 				Table: scanOpDef.Table,
> 				Index: i,
> 				// Grab what columns we can out of the secondary index. We'll get the
> ```

I believe we currently get all columns from the primary index. We only use the PK from the secondary.

In any case, it shouldn't matter how we get them when looking at the semantics of IndexJoinOp. I think `Cols` should have all columns we scan (one way or the other).

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/25019#-:-LAoQ2sg3rI2OYCwKjyJ:b-dvkohu)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
I think we should just call it `LookupJoin` to begin with. Is there some reason why that wouldn't work? It's OK that it's currently limited to just doing a lookup on the primary index. We'll add an `Index` ordinal field later on.

---

Review status: 0 of 17 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-:-LApCboT14YfiwAN44Qn:b15r9ie)*
<!-- Sent from Reviewable.io -->

justinj commented:
Just making it `LookupJoin` is fine with me, but are we consolidating index joins and lookup joins on the execution side? Those are two separate processors today in DistSQL, aren't they? Local SQL doesn't even have generic lookup joins, so we might have to special-case creating an `indexJoinNode` at execbuild time in its case.

petermattis commented:
> Local SQL doesn't even have generic lookup joins, so we might have to special-case creating an indexJoinNode at execbuild time in its case.

While this is currently true (and applies to merge joins as well), the work to merge the two execution engines in 2.1 will remove this limitation.

RaduBerinde commented:
Yeah we might do some special casing in the execbuilder for now, but that's fairly isolated and preferable to having different Ops. In distsql there is only one processor (JoinReader), though it is configured differently for the two cases (but we have an open issue to remove that differentiation).

rytaft commented:
Reviewed 17 of 17 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/opt/memo/private_defs_test.go, line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAsARSw8q4_2HbngSWJ:-LAsARSw8q4_2HbngSWK:bszz4yo) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/memo/private_defs_test.go#L42)):*
> ```Go
> 	test := func(def *memo.ScanOpDef, altIndex int, expected bool) {
> 		t.Helper()
> 		if def.AltIndexMissingCols(md, altIndex).Empty() != expected {
> ```

Now that you're no longer returning a boolean I would test that it returns the expected ColSet.

---

*[pkg/sql/opt/norm/testdata/combo, line 320 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAsCYm11QMPEA0lhIeq:-LAsCYm11QMPEA0lhIer:b859thh) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/norm/testdata/combo#L320)):*
> ```
>   + └── columns: s:4(string)
> --------------------------------------------------------------------------------
> ConstrainScan (higher cost)
> ```

Where is the scan constrained?

---

*[pkg/sql/opt/norm/testdata/combo, line 1267 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAsHmbs9wnwKvNkjiNP:-LAsCtJ294Pl8C-Riojy:bgw6dyc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/norm/testdata/combo#L1267)):*
> ```
>   + └── ordering: +2
> --------------------------------------------------------------------------------
> GenerateIndexScans (higher cost)
> ```

where is the index scan? (should we be able to see it?)

---

*[pkg/sql/opt/xform/coster.go, line 75 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAsFsWpA3PXnzV9sCvE:-LAsFsWpA3PXnzV9sCvF:b-2ajrle) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/xform/coster.go#L75)):*
> ```Go
> 	case opt.IndexJoinOp:
> 		// TODO(justin): remove this once we can execbuild index joins.
> 		return memo.MaxCost
> ```

I thought we only needed to ensure execbuild parity for scalar expressions (right, @RaduBerinde?)

I think it would be useful for testing to see a case when index join is chosen over a regular scan...

---

*[pkg/sql/opt/xform/physical_props_factory.go, line 87 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAsEzfL2wMCtOTzGiW5:-LAsEzfM7c-ImkQ8urdC:bnny9ta) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/xform/physical_props_factory.go#L87)):*
> ```Go
> 
> 	case opt.ProjectOp, opt.IndexJoinOp:
> 		// Operators which can pass through ordering if they operate only on input
> ```

[nit] to make this comment parallel with the one above, I'd say "Project and Index Join operators can pass through ordering if it operates only on input columns."

---

*[pkg/sql/opt/xform/testdata/rules/scan, line 146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAsGlw31_nLpKX0gysM:-LAsGlw31_nLpKX0gysN:bzi7e1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/xform/testdata/rules/scan#L146)):*
> ```
>  │         └── cost: 1000.00
>  ├── G3: (projections G7 G6)
>  ├── G4: (scan abc@bc,columns=())
> ```

why is this one generated if it has 0 columns?

---

*[pkg/sql/opt/xform/testdata/rules/scan, line 150 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAsGv8ABQZPcS0290x0:-LAsGv8ABQZPcS0290x1:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/xform/testdata/rules/scan#L150)):*
> ```
>  │         ├── best: (scan abc@bc,columns=())
>  │         └── cost: 1000.00
>  ├── G5: (scan abc@ba,columns=())
> ```

ditto

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25019)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
We will need exploration rules to push down filters through the indexjoin. For example:
  Table (K PRIMARY KEY, A, B, C)
  Index (C STORING A)
  Filter: WHERE C > 1 AND A > 5 AND B > 3

We would want this to be a constrained index scan, followed by the filter (A > 5), followed by the join, followed by the filter (B > 3).

---

Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/sql/opt/xform/coster.go, line 75 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAsFsWpA3PXnzV9sCvE:-LAsUmsO6MQpKwsqnKOz:b-a3ggya) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/xform/coster.go#L75)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I thought we only needed to ensure execbuild parity for scalar expressions (right, @RaduBerinde?)

I think it would be useful for testing to see a case when index join is chosen over a regular scan...
</blockquote></details>

Parity for scalar expressions is required for the index selection code which is used by the current planner.

Setting a high cost here until we can execute it just ensures we won't regress in terms of what queries the optimizer can run.

---

*[pkg/sql/opt/xform/explorer.go, line 243 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAoRPFQCH0EM29s9Qga:-LAsVWyL427uKZvYjYnr:bgtfl7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/xform/explorer.go#L243)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I believe we currently get all columns from the primary index. We only use the PK from the secondary.

In any case, it shouldn't matter how we get them when looking at the semantics of IndexJoinOp. I think `Cols` should have all columns we scan (one way or the other).
</blockquote></details>

Sorry I think I misread this code. This looks good, except that in practice we would only need the PKs and whatever columns might be needed for a filter that runs before the join. I don't quite know how to achieve that though.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-:-LAsWR6d4Qq3b9MyLz_w:b-p4kw46)*
<!-- Sent from Reviewable.io -->

petermattis commented:
> We will need exploration rules to push down filters through the indexjoin. For example:
Table (K PRIMARY KEY, A, B, C)
Index (C STORING A)
Filter: WHERE C > 1 AND A > 5 AND B > 3
> 
> We would want this to be a constrained index scan, followed by the filter (A > 5), followed by the join, followed by the filter (B > 3).

I was imagining we'd push filters onto scans before we determined that an index-join was necessary. But I don't see how that would work with the current structure.

andy-kimball commented:
How sophisticated is 2.0 with pushing down filters into a lookup join like this?

---

Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/sql/opt/xform/physical_props_factory.go, line 87 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAsEzfL2wMCtOTzGiW5:-LAsglYh8DykME6POjiR:b122rs8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/xform/physical_props_factory.go#L87)):*
Also, I'd clarify the comment a bit:
> Project and Index Join operators can pass through ordering if the ordering operates only on input columns.

---

*[pkg/sql/opt/xform/physical_props_factory.go, line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAshCN00yGVLAFEm3h0:-LAshCN00yGVLAFEm3h1:be3wbyo) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/xform/physical_props_factory.go#L89)):*
> ```Go
> 		// Operators which can pass through ordering if they operate only on input
> 		// columns.
> 		var input memo.GroupID
> ```

NIT: I think it'd be a bit better to make this code into a helper func, similar to:

`func (f physicalPropsFactory) orderingBoundBy(ordering memo.Ordering, input memo.GroupID) bool`

Then you can keep the `Project` and `IndexJoin` separate in the switch statement, rather than needing a nested switch.

---

*[pkg/sql/opt/xform/testdata/rules/limit, line 112 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAsiqH73su_SR-IdyId:-LAsiqH73su_SR-IdyIe:b533a9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/xform/testdata/rules/limit#L112)):*
> ```
> ----
> memo (optimized)
>  ├── G1: (limit G2 G3 ) (scan a@s_idx,columns=(4),constrained,lim=1) (scan a@si_idx,columns=(4),constrained,lim=1)
> ```

We should abbreviate to `cols`, since space is at a premium.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-:-LAsfK5a0fHA1aaWNSyZ:b-vappp5)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
2.0 does do this (the index scan node can have a filter)

---

Review status: all files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-:-LAsjv0ACQ24jAql_kvk:boiwidj)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
> I was imagining we'd push filters onto scans before we determined that an index-join was necessary. But I don't see how that would work with the current structure.

The trouble is that pushing filters into scans is different for each possible index, since indexes have different orderings and different available columns. The same filter can be pushed down in N different ways. This is why `Select` pushdown into `Scan` uses exploration rules. Similarly, pushing `Select` down through `LookupJoin` is going to require exploration rules.

petermattis commented:
> The trouble is that pushing filters into scans is different for each possible index, since indexes have different orderings and different available columns. The same filter can be pushed down in N different ways. This is why Select pushdown into Scan uses exploration rules. Similarly, pushing Select down through LookupJoin is going to require exploration rules.

Could an index-join be modeled as an inner-join that is later transformed into a lookup-join? Going through the inner-join state would allow the normal filter push down rules to apply. I'm guessing the challenge with this approach would be recognizing when the inner-join can be implemented as a lookup-join.

RaduBerinde commented:
> Could an index-join be modeled as an inner-join that is later transformed into a lookup-join? Going through the inner-join state would allow the normal filter push down rules to apply. I'm guessing the challenge with this approach would be recognizing when the inner-join can be implemented as a lookup-join.

Eventually, I'd think that's where we want to be. The generic logic for lookup join should just apply. Maybe with just the exception of taking care that the costs are such that we always prefer lookup in the index join case (even in the absence of stats).

andy-kimball commented:
> Could an index-join be modeled as an inner-join that is later transformed into a lookup-join? Going through the inner-join state would allow the normal filter push down rules to apply. I'm guessing the challenge with this approach would be recognizing when the inner-join can be implemented as a lookup-join.

The trouble is there are N different possible inner-joins to consider. For example, if index A has cols (1, 3, 4) and index B has cols (2, 3, 4), then there are two different possible inner joins. These kinds of combinations fall within the domain of exploration rules. Trying to make normalization rules handle this will likely run into a bunch of issues. For example, the inner-join in this case is really a self-join between a secondary index and a primary index. But to feed them into an inner-join, we must have different TableID's and ColumnID's. But then the optimizer would treat them as entirely different tables, and the filter push down rules wouldn't work as intended anyway.

justinj commented:
I've left out exploration/pushdown rules for the time being, could do those in this PR but it might be simpler to leave them for a follow-up.

---

Review status: 0 of 19 files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/sql/opt/memo/private_defs_test.go, line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAsARSw8q4_2HbngSWJ:-LAsqoMM-ndGpE9KO2tT:bumg0b9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/memo/private_defs_test.go#L42)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Now that you're no longer returning a boolean I would test that it returns the expected ColSet.
</blockquote></details>

I ended up changing this back.

---

*[pkg/sql/opt/norm/testdata/combo, line 317 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAoQMaF3aGvmbKZDKYW:-LAsp2l28zDJmKbHXPXq:b-cw76cz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/norm/testdata/combo#L317)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

unknown?
</blockquote></details>

Fixed, this was a problem with having this case be MaxCost, I've changed it to just be a high cost, now.

---

*[pkg/sql/opt/norm/testdata/combo, line 320 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAsCYm11QMPEA0lhIeq:-LAspHZx1k3nBmE_w_08:b-zidqhq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/norm/testdata/combo#L320)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Where is the scan constrained?
</blockquote></details>

I think this is fixed now.

---

*[pkg/sql/opt/norm/testdata/combo, line 1267 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAsHmbs9wnwKvNkjiNP:-LAspJ4GD0dHDhs_xLw8:b-4p6e7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/norm/testdata/combo#L1267)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

where is the index scan? (should we be able to see it?)
</blockquote></details>

This was a problem with me using MaxCost elsewhere and not overwriting the default placeholder - I've switched it to just use a high cost.

---

*[pkg/sql/opt/xform/explorer.go, line 243 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAoRPFQCH0EM29s9Qga:-LAs6HpkCYA_Vp7FSex4:b-7eo5fb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/xform/explorer.go#L243)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Sorry I think I misread this code. This looks good, except that in practice we would only need the PKs and whatever columns might be needed for a filter that runs before the join. I don't quite know how to achieve that though.
</blockquote></details>

Right, I think we should actually take *all* the columns from the secondary index here (since stuff might need to get pushed down into the scan), and then remove them later once they've been used (or not) for various filters and orderings? Does that seem reasonable?

---

*[pkg/sql/opt/xform/physical_props_factory.go, line 87 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAsEzfL2wMCtOTzGiW5:-LAssBX89ilBFf-j4BFY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/xform/physical_props_factory.go#L87)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Also, I'd clarify the comment a bit:
> Project and Index Join operators can pass through ordering if the ordering operates only on input columns.
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/physical_props_factory.go, line 89 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAshCN00yGVLAFEm3h0:-LAssCY2CJOM8HZCqgLf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/xform/physical_props_factory.go#L89)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: I think it'd be a bit better to make this code into a helper func, similar to:

`func (f physicalPropsFactory) orderingBoundBy(ordering memo.Ordering, input memo.GroupID) bool`

Then you can keep the `Project` and `IndexJoin` separate in the switch statement, rather than needing a nested switch.
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/testdata/rules/limit, line 112 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAsiqH73su_SR-IdyId:-LAssKEvD74H0_H3dfzT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/xform/testdata/rules/limit#L112)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

We should abbreviate to `cols`, since space is at a premium.
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/testdata/rules/scan, line 146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAsGlw31_nLpKX0gysM:-LAssvBpCoe8pefZmPTJ:b-7vnk9y) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/xform/testdata/rules/scan#L146)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

why is this one generated if it has 0 columns?
</blockquote></details>

Fixed, we properly get the number of columns now.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-:-LAsuWtr7lrtVOLOIsfg:b-g889jb)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 0 of 19 files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/sql/opt/xform/explorer.go, line 243 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAoRPFQCH0EM29s9Qga:-LAtB1nOBB9U6GDg14FI:bnei2bc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/xform/explorer.go#L243)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Right, I think we should actually take *all* the columns from the secondary index here (since stuff might need to get pushed down into the scan), and then remove them later once they've been used (or not) for various filters and orderings? Does that seem reasonable?
</blockquote></details>

I think this:

`((Scan Columns) UNION (Lookup Join Key Columns)) DIFF (Missing Columns)`

You don't need secondary index cols that aren't projected by the scan and aren't part of the join key.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25019)*
<!-- Sent from Reviewable.io -->

rytaft commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> nice!

---

Reviewed 19 of 19 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/sql/opt/metadata.go, line 108 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAtBEQGAs5TeV4Lmz_h:-LAtBEQGAs5TeV4Lmz_i:b-r7t7os) ([raw file](https://github.com/cockroachdb/cockroach/blob/9939a77aa38eb44544c21d44cbbf8b50bf3ac46c/pkg/sql/opt/metadata.go#L108)):*
> ```Go
> }
> 
> // IndexColumns returns the set of columns in some index.
> ```

[nit] I'd say "the given index" instead of "some index"

---

*[pkg/sql/opt/metadata.go, line 114 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAtBWli-WyPkBjHWROq:-LAtBWli-WyPkBjHWROr:bgnqey1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9939a77aa38eb44544c21d44cbbf8b50bf3ac46c/pkg/sql/opt/metadata.go#L114)):*
> ```Go
> 
> 	var indexCols ColSet
> 	for j := 0; j < index.ColumnCount(); j++ {
> ```

[nit] I'd use `i` instead of `j` to stick with convention

---

*[pkg/sql/opt/metadata_test.go, line 207 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAtCcPm3maTksYbiR7M:-LAtCcPm3maTksYbiR7N:bt38bya) ([raw file](https://github.com/cockroachdb/cockroach/blob/9939a77aa38eb44544c21d44cbbf8b50bf3ac46c/pkg/sql/opt/metadata_test.go#L207)):*
> ```Go
> 	s := int(md.TableColumn(a, 2))
> 
> 	test := func(def *memo.ScanOpDef, altIndex int, expected bool) {
> ```

Can you add a comment somewhere to explain what this test is doing?

---

*[pkg/sql/opt/norm/testdata/combo, line 320 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAsCYm11QMPEA0lhIeq:-LAtDWqr40bsu6yKI-jZ:b-f9us9u) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/norm/testdata/combo#L320)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I think this is fixed now.
</blockquote></details>

It's still not showing a constraint on the scan... but I think that's because `ConstrainScan` actually doesn't add any new expressions here. It's a bit odd that it's showing a diff, but I'm not sure if it's worth fixing or how difficult it would be to fix. @andy-kimball, do you have any thoughts about this?

---

*[pkg/sql/opt/xform/coster.go, line 75 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAsFsWpA3PXnzV9sCvE:-LAtFVuX2-TWyqGs9GBq:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/xform/coster.go#L75)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Parity for scalar expressions is required for the index selection code which is used by the current planner.

Setting a high cost here until we can execute it just ensures we won't regress in terms of what queries the optimizer can run.
</blockquote></details>

:thumbsup: 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-:-LAtGW7b5DGXOQMc73qW:b-7yytuz)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/sql/opt/norm/testdata/combo, line 1256 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAtPSEv0SaDbbRznM0o:-LAtPSEv0SaDbbRznM0p:b-l6p47x) ([raw file](https://github.com/cockroachdb/cockroach/blob/9939a77aa38eb44544c21d44cbbf8b50bf3ac46c/pkg/sql/opt/norm/testdata/combo#L1256)):*
> ```
>   + │    │    │         │    │    │    ├── keys: (1)
>   + │    │    │         │    │    │    └── scan a@secondary
>   + │    │    │         │    │    │         ├── columns: a.x:1(int!null) a.f:3(float) a.s:4(string) a.j:5(jsonb)
> ```

Related to my other comment: we don't want to be projecting the "f", "s", and "j" columns here because they're not needed as output columns or as join columns.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25019)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/sql/opt/metadata.go, line 109 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAw0Eab3xASiOYDXmLP:-LAw0Eab3xASiOYDXmLQ:b-5y5rlx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9939a77aa38eb44544c21d44cbbf8b50bf3ac46c/pkg/sql/opt/metadata.go#L109)):*
> ```Go
> 
> // IndexColumns returns the set of columns in some index.
> func (md *Metadata) IndexColumns(tableID TableID, indexOrdinal int) ColSet {
> ```

Add a TODO to cache the `ColSet`s in the table metadata, since we're probably going to repeatedly call this method for the same indexes during optimization.

---

*[pkg/sql/opt/norm/testdata/combo, line 320 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAsCYm11QMPEA0lhIeq:-LAw1Lvj2F-POaZCVFzu:b-1gq82q) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/norm/testdata/combo#L320)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

It's still not showing a constraint on the scan... but I think that's because `ConstrainScan` actually doesn't add any new expressions here. It's a bit odd that it's showing a diff, but I'm not sure if it's worth fixing or how difficult it would be to fix. @andy-kimball, do you have any thoughts about this?
</blockquote></details>

Hmm, i believe i can fix this. In general, we do the best we can to give a diff that is helpful, but which diff we show is the result of heuristics that don't always work perfectly. I'll see what I can do (in a different PR, of course).

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25019)*
<!-- Sent from Reviewable.io -->

justinj commented:
Review status: 14 of 19 files reviewed at latest revision, 12 unresolved discussions.

---

*[pkg/sql/opt/metadata.go, line 108 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAtBEQGAs5TeV4Lmz_h:-LAtJjq97tjzsdajyWm4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9939a77aa38eb44544c21d44cbbf8b50bf3ac46c/pkg/sql/opt/metadata.go#L108)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] I'd say "the given index" instead of "some index"
</blockquote></details>

Done.

---

*[pkg/sql/opt/metadata.go, line 109 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAw0Eab3xASiOYDXmLP:-LAx6SKB8x9Vf1O2JyQZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9939a77aa38eb44544c21d44cbbf8b50bf3ac46c/pkg/sql/opt/metadata.go#L109)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Add a TODO to cache the `ColSet`s in the table metadata, since we're probably going to repeatedly call this method for the same indexes during optimization.
</blockquote></details>

Done.

---

*[pkg/sql/opt/metadata.go, line 114 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAtBWli-WyPkBjHWROq:-LAtJmUm1quhi9e081hV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9939a77aa38eb44544c21d44cbbf8b50bf3ac46c/pkg/sql/opt/metadata.go#L114)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] I'd use `i` instead of `j` to stick with convention
</blockquote></details>

Done.

---

*[pkg/sql/opt/metadata_test.go, line 207 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAtCcPm3maTksYbiR7M:-LAtKDGJ4zwdnGNEyDE_:b-pvuzr7) ([raw file](https://github.com/cockroachdb/cockroach/blob/9939a77aa38eb44544c21d44cbbf8b50bf3ac46c/pkg/sql/opt/metadata_test.go#L207)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Can you add a comment somewhere to explain what this test is doing?
</blockquote></details>

Done. It was a direct translation of an old test, but I guess it makes more sense at this point to clean it up for this purpose.

---

*[pkg/sql/opt/norm/testdata/combo, line 1256 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAtPSEv0SaDbbRznM0o:-LAx6VpL3JLW85rmmfCx:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/9939a77aa38eb44544c21d44cbbf8b50bf3ac46c/pkg/sql/opt/norm/testdata/combo#L1256)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Related to my other comment: we don't want to be projecting the "f", "s", and "j" columns here because they're not needed as output columns or as join columns.
</blockquote></details>

Fixed.

---

*[pkg/sql/opt/xform/explorer.go, line 243 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAoRPFQCH0EM29s9Qga:-LAx6uF63EssEy5jRNqM:b-qg1apf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/xform/explorer.go#L243)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think this:

`((Scan Columns) UNION (Lookup Join Key Columns)) DIFF (Missing Columns)`

You don't need secondary index cols that aren't projected by the scan and aren't part of the join key.
</blockquote></details>

Done, I wrote it as `(index-cols ∩ scan-columns) ∪ pk-columns` which is the same in this case, but will have to be generalized when we actually have a variable lookup join key column.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25019)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, good work

---

Review status: 14 of 19 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-:-LAxKptf7C8AjLZK9lBN:bxo1df3)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/opt/norm/testdata/combo, line 320 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25019#-LAsCYm11QMPEA0lhIeq:-LAxLBT61a6RYpjlVddT:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a43cb9ab55090e9449436aa99f453be713ec4ce4/pkg/sql/opt/norm/testdata/combo#L320)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Hmm, i believe i can fix this. In general, we do the best we can to give a diff that is helpful, but which diff we show is the result of heuristics that don't always work perfectly. I'll see what I can do (in a different PR, of course).
</blockquote></details>

:thumbsup: 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25019)*
<!-- Sent from Reviewable.io -->

justinj commented:
TFTRs!

bors r+
